.NET 10 Preview 1 - WinForms Release Notes (#9722)
dotnet/core: release-notes/10.0/preview/preview1/winforms.md

1. *Replace obsolete APIs*  Start by replacing the obsolete `GetData` methods with the new `TryGetData<T>`. This ensures type safety and avoids the risks associated with unbounded deserialization.

> 1. *Use JSON Serialization:* For complex types, use JSON serialization instead of BinaryFormatter, replace `SetData` with the new APIs like `SetDataAsJson`.  The `TryGetData<T>` APIs are designed to handle common .NET types and JSON serialization.

Reviewer: ```suggestion
1. *Use JSON Serialization:* For complex types, use JSON serialization instead of BinaryFormatter, replace `SetData` with the new APIs like `SetDataAsJson`. The `TryGetData<T>` APIs are designed to handle common .NET types and JSON serialization.

```

For in-depth discussion of the APIs involved, and usage examples, please see the original [API Proposal](https://github.com/dotnet/winforms/issues/12362) in the WinForms repo.

### Obsoleted APIs

In .NET 10, WinForms has marked as obsolete the following APIs to help warn developers that BinaryFormatter may be used to deserialize data.

**Clipboard:**

- `public static object? GetData(string format)`

**DataObject:**

- `public virtual object? GetData(string format)`
- `public virtual object? GetData(string format, bool autoConvert)`
- `public virtual object? GetData(Type format)`

**ClipboardProxy (VisualBasic wrapper for WinForms Clipboard):**

- `public object GetData(string format)`

### New APIs

We’ve introduced the following APIs in Windows Forms for .NET 10 Preview 1:

**Clipboard:**

- `public static void SetDataAsJson<T>(string format, T data)`
- `public static bool TryGetData<T>(string format, out T data)`
- `public static bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, out T data)`

**IDataObject:**
**ITypedDataObject:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`

> - `public void SetDataAsJson<T>(string format, bool autoConvert, T data)`

Reviewer: - `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public void SetDataAsJson<T>(string format, bool autoConvert, T data)`
- `protected virtual bool TryGetDataCore<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`
These are not interface methods, they are added on the concrete DataObject class

Suggestions:
**DataObject:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public void SetDataAsJson<T>(string format, bool autoConvert, T data)`
- `public bool TryGetData<T>(out T data);`
- `public bool TryGetData<T>(string format, out T data);`
- `public bool TryGetData<T>(string format, bool autoConvert, out T data);`
- `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`
- `protected virtual bool TryGetDataCore<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`

**ITypedDataObject:**

- `public bool TryGetData<T>(out T data);`
- `public bool TryGetData<T>(string format, out T data);`
- `public bool TryGetData<T>(string format, bool autoConvert, out T data);`
- `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`

- `public bool TryGetData<T>(string format, bool autoConvert, out T data);`
- `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`
- `protected virtual bool TryGetDataCore<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`

**DataObjectExtensions:**

- `public static bool TryGetData<T>(this IDataObject dataObject, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, bool autoConvert, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`

**ClipboardProxy:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public bool TryGetData<T>(string format, out T data)`
- `public bool TryGetData<T>(string format, System.Func<System.Reflection.Metadata.TypeName, System.Type> resolver, out T data)`

### New Configuration Switch

> If an application is unable to adopt the new APIs, and requires a fallback to `BinaryFormatter` serialization during Clipboard and Drag/Drop operations, the developer will need to take additional action. In .NET 9, calls to BinaryFormatter will always throw a `PlatformNotSupportedException` unless the developer follows the instructions located here: [BinaryFormatter compatibility package](https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package) . In .NET 10 Preview 1, WinForms will now also by default disable the use of BinaryFormatter as a fallback unless it is explicitly enabled with this new configuration switch. This switch is `Windows.ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization`, and it must be set to `True` for the fallback to happen successfully.

Reviewer: typo: extra white space after the closing paren.

Reviewer: `In .NET 10 Preview 1, WinForms will now also by default disable`  ->> `In .NET 10 Preview 1, WinForms will also by default disable`

Author: Done, thanks!

# Windows Forms in .NET 10 Preview 1 - Release Notes

> WinForms continues to ease the pain of migrating away from BinaryFormatter in .NET 10 Preview 1 by simplifying Clipboard serialization for certain custom types. In .NET 9, calls to BinaryFormatter to perform serialization were updated to throw `PlatformNotSupportedException` if the developer did not reference an unsupported compatibility package and enable BinaryFormatter via an application configuration switch([see more](https://github.com/dotnet/winforms/issues/12362)). Clipboard and Drag/Drop scenarios which use custom types, as of .NET 9, require the developer to either enable BinaryFormatter or explicitly take action to use JSON serialization (or similar) to succeed. In .NET 10 Preview 1, WinForms is introducing APIs to help simplify the process of JSON serialization in Clipboard and Drag/Drop scenarios and obsoleting existing APIs to continue to encourage developers to move away from the unsafe BinaryFormatter serialization. For cases where BinaryFormatter is still required, a configuration switch has been added.

Reviewer 2: We should probably link to the official docs for the compat package and enabling the switch instead of an issue.
```suggestion
WinForms continues to ease the pain of migrating away from BinaryFormatter in .NET 10 Preview 1 by simplifying Clipboard serialization for certain custom types. In .NET 9, calls to BinaryFormatter to perform serialization were updated to throw `PlatformNotSupportedException` if the developer did not reference an unsupported compatibility package and enable BinaryFormatter via an application configuration switch([see more](https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package)). Clipboard and Drag/Drop scenarios which use custom types, as of .NET 9, require the developer to either enable BinaryFormatter or explicitly take action to use JSON serialization (or similar) to succeed. In .NET 10 Preview 1, WinForms is introducing APIs to help simplify the process of JSON serialization in Clipboard and Drag/Drop scenarios and obsoleting existing APIs to continue to encourage developers to move away from the unsafe BinaryFormatter serialization. For cases where BinaryFormatter is still required, a configuration switch has been added.

```

Author: Done, found some other silly mistakes too.

For in-depth discussion of the APIs involved, and usage examples, please see the original [API Proposal](https://github.com/dotnet/winforms/issues/12362) in the WinForms repo.

### Obsoleted APIs

> In .NET 10, WinForms has marked as obsolete the following APIs to help warn developers that BinaryFormatter may be used to deserialize data. For detailed usage information, please see the API proposal located *An external link was removed to protect your privacy.*.

Reviewer 2: It looks like there was supposed to be a link at the end of this

Author: Oops, I didn't want that sentence at all - I'd referred to the API Proposal already. Thanks!

- `public object GetData(string format)`

### New APIs

We’ve introduced the following APIs:

**Clipboard:**

- `public void SetDataAsJson<T>(T data) { }`
- `public void SetDataAsJson<T>(string format, T data) { }`
- `public void SetDataAsJson<T>(string format, bool autoConvert, T data) { }`
- `public virtual bool TryGetData<T>(out T data) { }`
- `public virtual bool TryGetData<T>(string format, out T data) { }`
- `public virtual bool TryGetData<T>(string format, bool autoConvert, out T data) { }`
- `public virtual bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data) { }`

**IDataObject:**

- `bool TryGetData<T>(out T data);`
- `bool TryGetData<T>(string format, out T data);`
- `bool TryGetData<T>(string format, bool autoConvert, out T data);`
- `bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`

**Control:**

- `public DragDropEffects DoDragDropAsJson<T>(T data, DragDropEffects allowedEffects, Bitmap? dragImage, Point cursorOffset, bool useDefaultDragImage)`
- `public DragDropEffects DoDragDropAsJson<T>(T data, DragDropEffects allowedEffects)`

> **DragDrop:**

Reviewer: This is WPF and this one is not in yet

- `public static DragDropEffects DoDragDropAsJson<T>(DependencyObject dragSource, T data, DragDropEffects allowedEffects)`

**ClipboardProxy:**

- `public void SetDataAsJson<T>(T data) { }`
- `public void SetDataAsJson<T>(string format, T data) { }`
- `public bool TryGetData<T>(string format, out T data) { }`
- `public bool TryGetData<T>(string format, System.Func<System.Reflection.Metadata.TypeName, System.Type> resolver, out T data) { }`

### New Configuration Switch

> If an application is unable to adopt the new APIs, and requires a fallback to BinaryFormatter serialization during Clipboard and Drag/Drop operations, the developer will need to take additional action. In .NET 9, calls to BinaryFormatter will always throw a `PlatformNotSupportedException` unless the developer follows the instructions *An external link was removed to protect your privacy.*. In .NET 10 Preview 1, WinForms will now also by default disable the use of BinaryFormatter as a fallback unless it is explicitly enabled with this new configuration switch. This switch is `ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization`, and it must be set to `True` for the fallback to happen successfully.

Reviewer 2: Ditto here with external link. I think we may want to include this one: https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package

**Clipboard:**

- `public object GetData(string format)`

**DataObject:**

- `public virtual object? GetData(string format)`
- `public virtual object? GetData(string format, bool autoConvert)`
- `public virtual object? GetData(Type format)`

**ClipboardProxy (VisualBasic wrapper for WinForms Clipboard):**

- `public object GetData(string format)`

### New APIs

We’ve introduced the following APIs:

**Clipboard:**

- `public void SetDataAsJson<T>(T data) { }`
- `public void SetDataAsJson<T>(string format, T data) { }`
- `public void SetDataAsJson<T>(string format, bool autoConvert, T data) { }`
- `public virtual bool TryGetData<T>(out T data) { }`
- `public virtual bool TryGetData<T>(string format, out T data) { }`
- `public virtual bool TryGetData<T>(string format, bool autoConvert, out T data) { }`
- `public virtual bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data) { }`

> **IDataObject:**

Reviewer: ```suggestion
**ITypedDataObject:**

```

Reviewer: @merriemcgaw  - see Jeremy Barton's recommendations summary he added to the API proposal bug after the meeting - he made some modifications to the original proposal

# Windows Forms in .NET 10 Preview 1 - Release Notes

> WinForms continues to ease the pain of migrating away from BinaryFormatter in .NET 10 Preview 1 by simplifying Clipboard serialization for certain custom types. In .NET 9, calls to BinaryFormatter to perform serialization were updated to throw `PlatformNotSupportedException` if the developer did not reference an unsupported compatibility package and enable BinaryFormatter via an application configuration switch ([see more](https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package)). Clipboard and Drag/Drop scenarios which use custom types, as of .NET 9, require the developer to either enable BinaryFormatter or explicitly take action to use JSON serialization (or similar) to succeed. For additional details about migrating WinForms application away from `BinaryFormatter` see the [Windows Forms migration guide for BinaryFormatter](https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/winforms-applications). In .NET 10 Preview 1 WinForms is adding several APIs to enable developers to more easily use JSON serliaization in while using the Clipboard. Additionally, developers can now more easily specify expected types when getting data from the Clipboard.

Reviewer: serialization and deserialization?
Serailization is safe. But people understand that both directions are implied?

Reviewer: `take action` -->> `serialize to and from a stream`

**Clipboard:**

- `public static object? GetData(string format)`

**DataObject:**

- `public virtual object? GetData(string format)`
- `public virtual object? GetData(string format, bool autoConvert)`
- `public virtual object? GetData(Type format)`

**ClipboardProxy (VisualBasic wrapper for WinForms Clipboard):**

- `public object GetData(string format)`

## New APIs

We’ve introduced the following APIs in Windows Forms for .NET 10 Preview 1:

**Clipboard:**

- `public static void SetDataAsJson<T>(string format, T data)`
- `public static bool TryGetData<T>(string format, out T data)`
- `public static bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, out T data)`

**DataObject:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public void SetDataAsJson<T>(string format, bool autoConvert, T data)`
- `public bool TryGetData<T>(out T data);`
- `public bool TryGetData<T>(string format, out T data);`
- `public bool TryGetData<T>(string format, bool autoConvert, out T data);`

> - `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`

Reviewer: Do we have to give the full namespace name? `Reflection.Metadata.TypeName` it's less readable this way, if people want details they will dig further.

Author: Updated

**ITypedDataObject:**

- `public bool TryGetData<T>(out T data);`
- `public bool TryGetData<T>(string format, out T data);`
- `public bool TryGetData<T>(string format, bool autoConvert, out T data);`
- `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`

**DataObjectExtensions:**

- `public static bool TryGetData<T>(this IDataObject dataObject, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, bool autoConvert, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`

**ClipboardProxy:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public bool TryGetData<T>(string format, out T data)`
- `public bool TryGetData<T>(string format, System.Func<System.Reflection.Metadata.TypeName, System.Type> resolver, out T data)`

## New Configuration Switch

> If an application is unable to adopt the new APIs, and requires a fallback to `BinaryFormatter` serialization during Clipboard and Drag/Drop operations, the developer will need to take additional action. In .NET 9, calls to `BinaryFormatter` will always throw a `PlatformNotSupportedException` unless the developer follows the instructions located here: [BinaryFormatter compatibility package](https://learn.microsoft.com/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package) . In .NET 10 Preview 1, WinForms will also by default disable the use of `BinaryFormatter` at the application level using the same compatibility switches. Additionally, we allow developer to control `BinaryFormatter` usage in Clipboard only with a new compatibility switch `Windows.ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization`. For full compatibility application should reference the unsupported compatibility package and opt in to both configuration switches. 

Reviewer: Only SetDataAsJson/TryGetData roundtrip does not require the BF. If users do SetData/TryGetData they still need BF, the advantage is that they are now following the best known practices and can opt-out from the BF in this particular scenario while keeping it for other logic in the app.

- `public bool TryGetData<T>(string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data);`

**DataObjectExtensions:**

- `public static bool TryGetData<T>(this IDataObject dataObject, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, bool autoConvert, out T data)`
- `public static bool TryGetData<T>(this IDataObject dataObject, string format, Func<Reflection.Metadata.TypeName, Type> resolver, bool autoConvert, out T data)`

**ClipboardProxy:**

- `public void SetDataAsJson<T>(T data)`
- `public void SetDataAsJson<T>(string format, T data)`
- `public bool TryGetData<T>(string format, out T data)`
- `public bool TryGetData<T>(string format, System.Func<System.Reflection.Metadata.TypeName, System.Type> resolver, out T data)`

## New Configuration Switch

- `Windows.ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization`

## Considerations for BinaryFormatter usage with the Clipboard

### Scenarios that work without any added reference to BinaryFormatter

> Applications that use the new .NET Clipboard APIs for standard data types or explicitly use JSON serialization will work seamlessly without needing any reference to `BinaryFormatter`. For instance, the new APIs like `SetDataAsJson` and `TryGetData<T>` are designed to handle common .NET types such as primitive types and `System.Drawing.Bitmap` without relying on `BinaryFormatter`. These APIs ensure type safety and provide mechanisms to constrain incoming data, making them suitable for most scenarios where data exchange is straightforward and does not involve complex types. Serialization using these methods is considered best practice, and developers are strongly encouraged to move in this direction.

Reviewer: ```suggestion
Applications that use the new .NET Clipboard APIs for standard data types or explicitly use JSON serialization will work without needing any reference to `BinaryFormatter`. For instance, the new APIs like `SetDataAsJson` and `TryGetData<T>` are designed to handle common .NET types such as primitive types and `System.Drawing.Bitmap` without relying on `BinaryFormatter`. These APIs ensure type safety and provide mechanisms to constrain incoming data, making them suitable for most scenarios where data exchange is straightforward and does not involve complex types. Serialization using these methods is considered best practice, and developers are strongly encouraged to move in this direction.

```

I removed `seamlessly` - they do have to rebuild and use the new APIs

### Scenarios that require full backwards compatibility in serialization

> For applications that still rely on `BinaryFormatter` for custom format serialization or need to maintain complete compatibility with .NET Framework applications, a reference to the out-of-band, unsupported `System.Runtime.Serialization.Formatters` package and the corresponding AppContext switch is required as detailed [here](https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package) . This is particularly necessary for scenarios where complex types are involved, and the application cannot migrate to the new JSON-based APIs. In such cases, developers must explicitly opt-in to enable BinaryFormatter serialization and deserialization by setting the `ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization` switch to `true` in the configuration file in addition to `EnableUnsafeBinaryFormatterSerialization` flag to `true`. The advantage to this approach is that `BinaryFormatter` can be explicitly disabled for Clipboard scenarios, while still being allowed in other parts of the application. This ensures that the application can continue to function while gradually transitioning to safer alternatives.

Reviewer: ```suggestion
For applications that still rely on `BinaryFormatter` for custom format serialization or need to maintain complete compatibility with .NET Framework applications, a reference to the out-of-band, unsupported `System.Runtime.Serialization.Formatters` package and the corresponding AppContext switch is required as detailed [here](https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-migration-guide/compatibility-package). This is particularly necessary for scenarios where complex types are involved, and the application cannot migrate to the new JSON-based APIs. In such cases, developers must explicitly opt-in to enable BinaryFormatter serialization and deserialization by setting the `Windows.ClipboardDragDrop.EnableUnsafeBinaryFormatterSerialization` switch to `true` in the configuration file in addition to `System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization` switch to `true`. The advantage to this approach is that `BinaryFormatter` can be explicitly disabled for Clipboard scenarios, while still being allowed in other parts of the application. This ensures that the application can continue to function while gradually transitioning to safer alternatives.

```

### Best practices to consider when migrating to the new Clipboard APIs

> 1. *Replace obsolete APIs*  Start by replacing the obsolete `GetData` and `SetData` methods with the new `TryGetData<T>` and `SetDataAsJson` methods. This ensures type safety and avoids the risks associated with unbounded deserialization.

Reviewer: ```suggestion
1. *Replace obsolete APIs*  Start by replacing the obsolete `GetData` methods with the new `TryGetData<T>`. This ensures type safety and avoids the risks associated with unbounded deserialization.

```

SetData are not obsolete - setting is always safe and we cover it in the next point

> 1. *Use JSON Serialization:* For complex types, use JSON serialization instead of BinaryFormatter. The new APIs like `SetDataAsJson` and `TryGetData<T>` are designed to handle common .NET types and JSON serialization.

Reviewer: ```suggestion
1. *Use JSON Serialization:* For complex types, use JSON serialization instead of BinaryFormatter, replace `SetData` with the new APIs like `SetDataAsJson`.  The `TryGetData<T>` APIs are designed to handle common .NET types and JSON serialization.

```

> 1. *Constrain incoming data:* The new APIs provide ways to ensure that only expected types are deserialized, reducing the risk of deserialization attacks.

Reviewer: We need to add a bullet point about using POCO types for data transfer

Reviewer: 1. *Use simple types for data exchange* Re-design your application to use simple types that have no inheritance hierarchies and have only primitive types as data fields.

> 1. *Opt-In only for compatibility:* If your application still relies on `BinaryFormatter` for custom format serialization or needs to maintain compatibility with .NET Framework applications, use the configuration switch to enable BinaryFormatter serialization and deserialization. This should be done cautiously and only when necessary. Separately enable BinaryFormatter support during Clipboard serialization and dersialization with the new configuration switch.

Reviewer: ```suggestion
1. *Opt-In only for compatibility:* If your application still relies on `BinaryFormatter` for custom format serialization or needs to maintain compatibility with .NET Framework applications, use the configuration switch to enable BinaryFormatter serialization and deserialization. This should be done cautiously and only when necessary. Separately enable BinaryFormatter support during Clipboard serialization and dersialization with the new configuration switch.
1. *Use simple types for data exchange:* Re-design your application to use simple types that have no inheritance hierarchies and have only primitive types as data fields.
```

1. *Replace obsolete APIs*  Start by replacing the obsolete `GetData` methods with the new `TryGetData<T>`. This ensures type safety and avoids the risks associated with unbounded deserialization.

> 1. *Use JSON Serialization:* For complex types, use JSON serialization instead of BinaryFormatter, replace `SetData` with the new APIs like `SetDataAsJson`. The `TryGetData<T>` APIs are designed to handle common .NET types and JSON serialization.

Reviewer 2: This looks like duplicated content

Author: It is a bit repetitive, let me tweak the summary paragraph to not be so detailed since the list covers it.
